README.md
# py-fpff

[![PyPi version](https://img.shields.io/pypi/v/py-fpff)](https://pypi.org/project/py-fpff/)
[![RTD build](https://img.shields.io/readthedocs/py-fpff)](https://py-fpff.readthedocs.io/en/latest/)

Library for working with FPFF.

## Getting Started

### Install
```
pip install py-fpff
```

### Sample Code
```
from py_fpff import FPFF, SectionType

with open('./input.fpff') as f:
    fpff = FPFF(f)
    fpff.append(SectionType.ASCII, 'Hello, world!')
    fpff.export('./exported')
```

## Testing

```
python -m unittest
```

## Building Documentation

```
# Install Sphinx and ReadTheDocs theme
pip install Sphinx sphinx_rtd_theme

cd ./docs
make html
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2021-08-15

### Added
- FPFF section insertion.
- Sphinx documentation.
- Changelog.
- Tests for empty, non-media, and media FPFF creation and section insert/remove.

### Changed
- Re-implemented core FPFF module.
- FileType renamed to SectionType.
- More verbose Python docstrings.
- Style improvements.

## [0.0.1] - 2018-06-06

### Added
- Initial FPFF implementation.

tests/test_fpff.py
import unittest
import os
import tempfile
import shutil
from py_fpff import FPFF, SectionType


class FPFFTest(unittest.TestCase):
    """FPFF test cases.
    """

    def setUp(self):
        self.test_dir = tempfile.mkdtemp()

    def tearDown(self):
        shutil.rmtree(self.test_dir)

    def test_empty(self):
        file_path = os.path.join(self.test_dir, 'out.fpff')

        fpff_1 = FPFF()
        with open(file_path, 'wb') as f:
            fpff_1.write(f)

        with open(file_path, 'rb') as f:
            fpff_2 = FPFF(f)
            assert fpff_2.version == 1
            assert fpff_2.author == ''
            assert fpff_2.nsects == 0

    def test_empty_2(self):
        file_path = os.path.join(self.test_dir, 'out.fpff')

        fpff_1 = FPFF(author='jasmaa')
        with open(file_path, 'wb') as f:
            fpff_1.write(f)

        with open(file_path, 'rb') as f:
            fpff_2 = FPFF(f)
            assert fpff_2.version == 1
            assert fpff_2.author == 'jasmaa'
            assert fpff_2.nsects == 0

    def test_nonmedia(self):
        file_path = os.path.join(self.test_dir, 'out.fpff')

        fpff_1 = FPFF(author='jasmaa')
        data = [
            (SectionType.ASCII, 'Hello, world!'),
            (SectionType.UTF8, 'おはよう世界'),
            (SectionType.WORDS, [b'\x00\x00\x00\x00', b'\xFF\xFF\xFF\xFF']),
            (SectionType.DWORDS, [
             b'\xFF\xFF\xFF\xFF\xFF\x9A\x00\xFF', b'\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF'
             ]),
            (SectionType.REF, 1),
            (SectionType.DOUBLES, [-1, 0, 0.3, -0.53]),
            (SectionType.COORD, (90.23, -200.34)),
        ]
        with open(file_path, 'wb') as f:
            for t, v in data:
                fpff_1.append(t, v)
            fpff_1.write(f)

        with open(file_path, 'rb') as f:
            fpff_2 = FPFF(f)
            assert fpff_2.version == 1
            assert fpff_2.author == 'jasmaa'
            assert fpff_2.nsects == len(data)

            cmp = list(zip(data, zip(fpff_2.stypes, fpff_2.svalues)))

            for (t1, v1), (t2, v2) in cmp[:5]:
                assert t1 == t2
                assert v1 == v2
            for (t1, v1), (t2, v2) in cmp[5:]:
                assert t1 == t2
                for n1, n2 in zip(v1, v2):
                    assert abs(n1 - n2) < 0.001

    def test_media(self):
        file_path = os.path.join(self.test_dir, 'out.fpff')

        with open(file_path, 'wb') as out_f:
            fpff_1 = FPFF(author='jasmaa')
            with open('./tests/test_png.png', 'rb') as in_f:
                data = in_f.read()
                fpff_1.append(SectionType.PNG, data)
            with open('./tests/test_gif87.gif', 'rb') as in_f:
                data = in_f.read()
                fpff_1.append(SectionType.GIF87, data)
            with open('./tests/test_gif89.gif', 'rb') as in_f:
                data = in_f.read()
                fpff_1.append(SectionType.GIF89, data)
            fpff_1.write(out_f)

        with open(file_path, 'rb') as f:
            fpff_2 = FPFF(f)
            assert fpff_2.version == 1
            assert fpff_2.author == 'jasmaa'
            assert fpff_2.nsects == 3

            with open('./tests/test_png.png', 'rb') as in_f:
                data = in_f.read()
                assert SectionType.PNG == fpff_2.stypes[0]
                assert data == fpff_2.svalues[0]
            with open('./tests/test_gif87.gif', 'rb') as in_f:
                data = in_f.read()
                assert SectionType.GIF87 == fpff_2.stypes[1]
                assert data == fpff_2.svalues[1]
            with open('./tests/test_gif89.gif', 'rb') as in_f:
                data = in_f.read()
                assert SectionType.GIF89 == fpff_2.stypes[2]
                assert data == fpff_2.svalues[2]

    def test_append_remove(self):
        fpff = FPFF(author='jasmaa')
        fpff.append(SectionType.ASCII, 'a')
        fpff.append(SectionType.ASCII, 'c')
        fpff.append(SectionType.ASCII, 'd')
        fpff.insert(1, SectionType.ASCII, 'b')
        fpff.remove(3)

        assert fpff.nsects == 3

        vals = ['a', 'b', 'c']
        for v1, v2 in zip(vals, fpff.svalues):
            assert v1 == v2


if __name__ == '__main__':
    unittest.main()


py-fpff.py
import binascii
from enum import Enum
import time
import struct
import os, shutil

"""
FPFF file type enum
"""
class FileType(Enum):
    ASCII = 1
    UTF8 = 2
    WORDS = 3
    DWORDS = 4
    DOUBLES = 5
    COORD = 6
    REF = 7
    PNG = 8
    GIF87 = 9
    GIF89 = 10

"""
Contains FPFF functions
"""
class FPFF():
    @staticmethod
    def reverse_bytearray(s):
        rev = bytearray()
        for i in range(len(s)-1, -1, -1):
            rev.append(s[i])
        return rev

    @staticmethod
    def remove_padding(data):
        while data[0] == 0:
            data.pop(0)
        return data
    @staticmethod
    def add_padding(data, l):

        if len(data) > l:
            raise ValueError("Data too large to be padded!")
        
        while len(data) < l:
            data.insert(0, 0)
        return data
    
    def __init__(self, file=None, author=None):
        self.version = 1
        self.timestamp = None
        self.author = None
        self.sections = 0
        self.stypes = list()
        self.svalues = list()

        if file != None:
            self.read(file)

    """
    Read in FPFF
    """
    def read(self, file):
        with open(file, "rb") as f:
            data = bytearray(f.read())

        magic            = FPFF.reverse_bytearray(data[0:4])
        self.version     = int.from_bytes(data[4:8], "little")
        self.timestamp   = int.from_bytes(data[8:12], "little")
        self.author      = FPFF.remove_padding(FPFF.reverse_bytearray(data[12:20])).decode('ascii')
        self.sect_num    = int.from_bytes(data[20:24], "little")

        # checks
        if magic != b'\xbe\xfe\xda\xde':
            print("Wrong header")
            return
        if self.version != 1:
            print("Not version 1")
            return

        # read sections
        count = 24
        for i in range(self.sect_num):
            stype = int.from_bytes(data[count:count+4], "little")
            slen =  int.from_bytes(data[count+4:count+8], "little")
            count += 8
            svalue = data[count:count+slen]

            # ascii
            if stype == 1:
                self.stypes.append(FileType.ASCII)
                self.svalues.append(svalue.decode('ascii'))
            # utf-8
            elif stype == 2:
                self.stypes.append(FileType.UTF8)
                self.svalues.append(svalue.decode('utf8'))
            # words
            elif stype == 3:
                self.stypes.append(FileType.WORDS)
                self.svalues.append([bytes(svalue[j:j+4]) for j in range(0, slen, 4)])
            # dwords
            elif stype == 4:
                self.stypes.append(FileType.DWORDS)
                self.svalues.append([bytes(svalue[j:j+8]) for j in range(0, slen, 8)])
            # doubles
            elif stype == 5:
                self.stypes.append(FileType.DOUBLES)
                self.svalues.append([int.from_bytes(svalue[j:j+8], "big") for j in range(0, slen, 8)])
            # coord
            elif stype == 6:
                self.stypes.append(FileType.COORD)
                self.svalues.append( (int.from_bytes(svalue[0:8],"big"), int.from_bytes(svalue[8:16],"big")) )
            # ref
            elif stype == 7:
                self.stypes.append(FileType.REF)
                self.svalues.append(int.from_bytes(svalue[0:4], "big"))
            # png
            elif stype == 8:
                self.stypes.append(FileType.PNG)
                sig = b'\x89\x50\x4E\x47\x0D\x0A\x1A\x0A'
                out = sig + svalue[0:slen]
                self.svalues.append(out)
            #gif87a
            elif stype == 9:
                self.stypes.append(FileType.GIF87)                
                sig = b'\x47\x49\x46\x38\x37\x61'
                out = sig + svalue[0:slen]
                self.svalues.append(out)
            #gif89a
            elif stype == 10:
                self.stypes.append(FileType.GIF89)                
                sig = b'\x47\x49\x46\x38\x39\x61'
                out = sig + svalue[0:slen]
                self.svalues.append(out)
            
            count += slen
            

    """
    Write to FPFF file
    """
    def write(self, file):
        # convert to bytes
        w_magic       = bytearray(b'\xDE\xDA\xFE\xBE')
        w_version     = FPFF.reverse_bytearray(FPFF.add_padding(struct.pack(">I", self.version), 4))
        w_timestamp   = FPFF.reverse_bytearray(FPFF.add_padding(struct.pack(">I", int(time.time())), 4))
        w_author      = FPFF.reverse_bytearray(FPFF.add_padding(bytearray(self.author, 'ascii'), 8))
        w_sect_num    = FPFF.reverse_bytearray(FPFF.add_padding(struct.pack(">I", self.sect_num), 4))
        w_sections  = list()
        for i in range(self.sect_num):
            
            w_svalue = None

            if self.stypes[i] == FileType.ASCII:
                w_svalue = bytearray(self.svalues[i], 'ascii')
            elif self.stypes[i] == FileType.UTF8:
                w_svalue = bytearray(self.svalues[i], 'utf8')
            elif self.stypes[i] == FileType.WORDS:
                w_svalue = b''.join(self.svalues[i])
            elif self.stypes[i] == FileType.DWORDS:
                w_svalue = b''.join(self.svalues[i])
            elif self.stypes[i] == FileType.DOUBLES:
                w_svalue = bytearray()
                for b in self.svalues[i]:
                    w_svalue.extend(FPFF.add_padding(bytearray.fromhex(hex(b)[2:]), 8))
            elif self.stypes[i] == FileType.COORD:
                w_svalue = bytearray()
                w_svalue.extend(FPFF.add_padding(bytearray.fromhex(hex(self.svalues[i][0])[2:]), 8))
                w_svalue.extend(FPFF.add_padding(bytearray.fromhex(hex(self.svalues[i][1])[2:]), 8))
            elif self.stypes[i] == FileType.REF:
                w_svalue = FPFF.add_padding(bytearray.fromhex(hex(self.svalues[i])[2:]), 4)
            elif self.stypes[i] == FileType.PNG:
                w_svalue = bytearray(self.svalues[i])
                del w_svalue[:8]
            elif self.stypes[i] == FileType.GIF87:
                w_svalue = bytearray(self.svalues[i])
                del w_svalue[:6]
            elif self.stypes[i] == FileType.GIF89:
                w_svalue = bytearray(self.svalues[i])
                del w_svalue[:6]

            w_slen = len(w_svalue)
            w_section = bytearray()
            w_section.extend(FPFF.reverse_bytearray(FPFF.add_padding(struct.pack(">I", int(self.stypes[i].value)), 4)))
            w_section.extend(FPFF.reverse_bytearray(FPFF.add_padding(struct.pack(">I", w_slen), 4)))
            w_section.extend(w_svalue)

            w_sections.extend(w_section)

        # construct and write
        out_data = bytearray()
        out_data.extend(w_magic)
        out_data.extend(w_version)
        out_data.extend(w_timestamp)
        out_data.extend(w_author)
        out_data.extend(w_sect_num)
        out_data.extend(w_sections)
        with open(file, 'wb') as f:
            f.write(bytes(out_data))
            f.close()

    """
    Export FPFF data to folder
    """
    def export(self, path):

        # create path
        dirpath = "/".join(path.split('/')[:-1])
        dirname = path.split('/')[-1]

        if dirpath != "":
            dirpath += "/"

        if os.path.exists(dirpath+dirname+"-data"):
            shutil.rmtree(dirpath+dirname+"-data")  
        os.makedirs(dirpath+dirname+"-data")

        # export files
        for i in range(self.sect_num):
            out_name = dirpath+dirname+"-data/"+dirname+"-"+str(i+1)
            w_svalue = None

            if self.stypes[i] in [FileType.ASCII, FileType.UTF8, FileType.WORDS, FileType.DWORDS, FileType.DOUBLES, FileType.COORD, FileType.REF]:

                if self.stypes[i] == FileType.ASCII:
                    w_svalue = self.svalues[i]
                elif self.stypes[i] == FileType.UTF8:
                    w_svalue = self.svalues[i]
                elif self.stypes[i] == FileType.WORDS:
                    w_svalue = " ".join([ val.hex() for val in self.svalues[i]])
                elif self.stypes[i] == FileType.DWORDS:
                    w_svalue = " ".join([ val.hex() for val in self.svalues[i]])
                elif self.stypes[i] == FileType.DOUBLES:
                    w_svalue = " ".join([ str(val) for val in self.svalues[i]])
                elif self.stypes[i] == FileType.COORD:
                    w_svalue = "LAT: " + str(self.svalues[i][0]) + "\nLON: " + str(self.svalues[i][1])
                elif self.stypes[i] == FileType.REF:
                    w_svalue = "REF: " + str(self.svalues[i])
                
                with open(out_name+".txt", 'w') as f:
                    f.write(w_svalue)
                    f.close()
            else:
                if self.stypes[i] == FileType.PNG:
                    with open(out_name+".png", 'wb') as f:
                        f.write(self.svalues[i])
                        f.close()
                elif self.stypes[i] == FileType.GIF87:
                    with open(out_name+".gif", 'wb') as f:
                        f.write(self.svalues[i])
                        f.close()
                elif self.stypes[i] == FileType.GIF89:
                    with open(out_name+".gif", 'wb') as f:
                        f.write(self.svalues[i])
                        f.close()
        

    def __repr__(self):
        return str(self.stypes)

s = FPFF("testing/test.fpff")
s.export('testing/lol')


py_fpff.py
"""Provides operations for working with FPFF.
"""

import os
import shutil
import time
import struct
from typing import Any, BinaryIO, Union
from enum import IntEnum


class SectionType(IntEnum):
    """Enum types for FPFF section types.

    Example:
        >>> t = SectionType.ASCII
    """

    ASCII = 1
    UTF8 = 2
    WORDS = 3
    DWORDS = 4
    DOUBLES = 5
    COORD = 6
    REF = 7
    PNG = 8
    GIF87 = 9
    GIF89 = 10


class FPFF:
    """FPFF file.

    Attributes:
        version (int): FPFF version number. Only accepts 1.
        timestamp (int): UNIX timestamp indicating time of creation.
        author (str): Author name.
        nsects (int): Number of sections in the FPFF.
        stypes (List[SectionType]): List of section types indexed by section.
        svalues (List[Any]): List of section values indexed by section.
    """

    def __init__(self, file: Union[BinaryIO, None] = None, author: str = ''):
        """Initializes new FPFF.

        Args:
            file (BinaryIO): Optional FPFF input byte stream. Provided if reading FPFF.
            author (str): Author name. Defaults to ''.

        Example:
            >>> with open('./input.fpff', 'rb') as f:
            >>>     fpff = FPFF(f)
        """

        self.version = 1
        self.timestamp = int(time.time())
        self.author = author
        self.nsects = 0
        self.stypes = list()
        self.svalues = list()

        # Read FPFF file if supplied
        if file != None:
            self.read(file)

    def read(self, file: BinaryIO):
        """Reads in FPFF from byte stream.

        Args:
            file (BinaryIO): FPFF input byte stream.

        Raises:
            ValueError: Magic did not match FPFF magic.
            ValueError: Unsupported version. Only version 1 is supported.
            ValueError: Section length must be greater than 0.
            ValueError: Improper section length.
            ValueError: Reference value is out of bounds.
            ValueError: File contained an unsupported type.

        Example:
            >>> with open('./input.fpff', 'rb') as f:
            >>>     fpff = FPFF()
            >>>     fpff.read(f)
        """

        data = file.read(24)

        magic = data[0:4][::-1]
        self.version = int.from_bytes(data[4:8], 'little')
        self.timestamp = int.from_bytes(data[8:12], 'little')
        self.author = data[12:20][::-1].decode('ascii').strip('\0')
        self.nsects = int.from_bytes(data[20:24], 'little')
        self.stypes = []
        self.svalues = []

        # Metadata checks
        if magic != b'\xBE\xFE\xDA\xDE':
            raise ValueError("Magic did not match FPFF magic.")
        if self.version != 1:
            raise ValueError(
                "Unsupported version. Only version 1 is supported."
            )

        # Read each section
        for _ in range(self.nsects):
            # Read section header and data
            data = file.read(8)
            stype = int.from_bytes(data[0:4], 'little')
            slen = int.from_bytes(data[4:8], 'little')

            if slen <= 0:
                raise ValueError("Section length must be greater than 0.")

            svalue = file.read(slen)

            # Decode section data
            if stype == 0x1:
                # ASCII
                self.stypes.append(SectionType.ASCII)
                self.svalues.append(svalue.decode('ascii'))
            elif stype == 0x2:
                # UTF-8
                self.stypes.append(SectionType.UTF8)
                self.svalues.append(svalue.decode('utf8'))
            elif stype == 0x3:
                # Words
                if slen % 4 != 0:
                    raise ValueError("Improper section length.")
                self.stypes.append(SectionType.WORDS)
                self.svalues.append(
                    [svalue[j:j+4] for j in range(0, slen, 4)]
                )
            elif stype == 0x4:
                # DWords
                if slen % 8 != 0:
                    raise ValueError("Improper section length.")
                self.stypes.append(SectionType.DWORDS)
                self.svalues.append(
                    [svalue[j:j+8] for j in range(0, slen, 8)]
                )
            elif stype == 0x5:
                # Doubles
                if slen % 8 != 0:
                    raise ValueError("Improper section length.")
                self.stypes.append(SectionType.DOUBLES)
                self.svalues.append(
                    [struct.unpack("<d", svalue[j:j+8])[0]
                     for j in range(0, slen, 8)]
                )
            elif stype == 0x6:
                # Coord
                if slen != 16:
                    raise ValueError("Improper section length.")
                self.stypes.append(SectionType.COORD)
                lat = struct.unpack("<d", svalue[0:8])[0]
                lng = struct.unpack("<d", svalue[8:16])[0]
                # TODO: validate lat and lng
                self.svalues.append((lat, lng))
            elif stype == 0x7:
                # Reference
                if slen != 4:
                    raise ValueError("Improper section length.")
                ref = int.from_bytes(svalue[0:4], 'little')
                if ref < 0 or ref >= self.nsects:
                    raise ValueError("Reference value is out of bounds.")
                self.stypes.append(SectionType.REF)
                self.svalues.append(ref)
            elif stype == 0x8:
                # PNG
                self.stypes.append(SectionType.PNG)
                sig = b'\x89\x50\x4E\x47\x0D\x0A\x1A\x0A'
                out = sig + svalue[0:slen]
                self.svalues.append(out)
            elif stype == 0x9:
                # GIF87a
                self.stypes.append(SectionType.GIF87)
                sig = b'\x47\x49\x46\x38\x37\x61'
                out = sig + svalue[0:slen]
                self.svalues.append(out)
            elif stype == 0xA:
                # GIF89a
                self.stypes.append(SectionType.GIF89)
                sig = b'\x47\x49\x46\x38\x39\x61'
                out = sig + svalue[0:slen]
                self.svalues.append(out)
            else:
                raise ValueError("File contained an unsupported type.")

    def write(self, file: BinaryIO):
        """Writes FPFF to byte stream.

        Args:
            file (BinaryIO): FPFF output byte stream.

        Raises:
            ValueError: Word needs to be 4 bytes.
            ValueError: DWord needs to be 8 bytes.

        Example:
            >>> with open('./output.fpff', 'wb') as out_f:
            >>>     fpff = FPFF()
            >>>     fpff.append(SectionType.ASCII, 'Hello, world!')
            >>>     fpff.write(out_f)
        """

        # Write FPFF header
        file.write(b'\xDE\xDA\xFE\xBE')
        file.write(self.version.to_bytes(4, 'little'))
        file.write(self.timestamp.to_bytes(4, 'little'))
        author_bytes = self.author.encode('ascii')[::-1]
        file.write(author_bytes)
        file.write(b'\x00'*(8-len(author_bytes)))
        file.write(self.nsects.to_bytes(4, 'little'))

        # Write each section
        for i in range(self.nsects):
            section_bytes = b''

            if self.stypes[i] == SectionType.ASCII:
                # ASCII
                section_bytes = self.svalues[i].encode('ascii')
            elif self.stypes[i] == SectionType.UTF8:
                # UTF-8
                section_bytes = self.svalues[i].encode('utf8')
            elif self.stypes[i] == SectionType.WORDS:
                # Words
                for w in self.svalues[i]:
                    if len(w) != 4:
                        raise ValueError("Word needs to be 4 bytes.")
                    section_bytes += w
            elif self.stypes[i] == SectionType.DWORDS:
                # DWords
                for w in self.svalues[i]:
                    if len(w) != 8:
                        raise ValueError("DWord needs to be 8 bytes.")
                    section_bytes += w
            elif self.stypes[i] == SectionType.DOUBLES:
                # Doubles
                section_bytes = b''.join(
                    [struct.pack("<d", w) for w in self.svalues[i]]
                )
            elif self.stypes[i] == SectionType.COORD:
                # Coords
                section_bytes = struct.pack("<d", self.svalues[i][0])
                section_bytes += struct.pack("<d", self.svalues[i][1])
            elif self.stypes[i] == SectionType.REF:
                # Reference
                section_bytes = self.svalues[i].to_bytes(4, 'little')
            elif self.stypes[i] == SectionType.PNG:
                # PNG
                section_bytes = self.svalues[i][8:]
            elif self.stypes[i] == SectionType.GIF87:
                # GIF87a
                section_bytes = self.svalues[i][6:]
            elif self.stypes[i] == SectionType.GIF89:
                # GIF89a
                section_bytes = self.svalues[i][6:]

            # Write to file
            file.write(self.stypes[i].to_bytes(4, 'little'))
            file.write(len(section_bytes).to_bytes(4, 'little'))
            file.write(section_bytes)

    def export(self, output_path: str):
        """Exports FPFF sections to directory.

        Args:
            output_path (str): Path to export directory.

        Example:
            >>> with open('./input.fpff') as f:
            >>>     fpff = FPFF(f)
            >>>     fpff.export('./exported')
        """

        # Ensure output path exists
        if os.path.exists(output_path):
            shutil.rmtree(output_path)
        os.mkdir(output_path)

        # Export files
        for i in range(self.nsects):
            if self.stypes[i] not in [SectionType.PNG, SectionType.GIF87, SectionType.GIF89]:
                # Non-media section
                file_name = f'section-{i}.txt'
                output = ''
                if self.stypes[i] == SectionType.ASCII:
                    output = self.svalues[i]
                elif self.stypes[i] == SectionType.UTF8:
                    output = self.svalues[i]
                elif self.stypes[i] == SectionType.WORDS:
                    output = ', '.join(
                        [val.hex() for val in self.svalues[i]]
                    )
                elif self.stypes[i] == SectionType.DWORDS:
                    output = ', '.join(
                        [val.hex() for val in self.svalues[i]]
                    )
                elif self.stypes[i] == SectionType.DOUBLES:
                    output = ', '.join(
                        [str(val) for val in self.svalues[i]]
                    )
                elif self.stypes[i] == SectionType.COORD:
                    output = f'LAT: {str(self.svalues[i][0])}\nLNG: {str(self.svalues[i][1])}'
                elif self.stypes[i] == SectionType.REF:
                    output = f'REF: {str(self.svalues[i])}'

                with open(os.path.join(output_path, file_name), 'w', encoding='utf8') as f:
                    f.write(output)

            else:
                # Media section
                if self.stypes[i] == SectionType.PNG:
                    file_name = f'section-{i}.png'
                    with open(os.path.join(output_path, file_name), 'wb') as f:
                        f.write(self.svalues[i])
                elif self.stypes[i] == SectionType.GIF87:
                    file_name = f'section-{i}.gif'
                    with open(os.path.join(output_path, file_name), 'wb') as f:
                        f.write(self.svalues[i])
                elif self.stypes[i] == SectionType.GIF89:
                    file_name = f'section-{i}.gif'
                    with open(os.path.join(output_path, file_name), 'wb') as f:
                        f.write(self.svalues[i])

    def insert(self, section_idx: int, obj_type: SectionType, obj_data: Any):
        """Inserts section before indicated index.

        Args:
            section_idx (int): Section index to insert in front of.
            obj_type (int): Section type of new section.
            obj_data (Any): Section value of new section.

        Raises:
            TypeError: Object data not valid for object type.

        Example:
            >>> fpff.insert(0, SectionType.ASCII, 'Hello, world!')
        """

        if obj_type == SectionType.ASCII and type(obj_data) == str:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.UTF8 and type(obj_data) == str:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.WORDS and type(obj_data) == list:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.DWORDS and type(obj_data) == list:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.DOUBLES and type(obj_data) == list:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.COORD and type(obj_data) == tuple:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.REF and type(obj_data) == int:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.PNG and type(obj_data) in [bytes, bytearray]:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.GIF87 and type(obj_data) in [bytes, bytearray]:
            self.svalues.insert(section_idx, obj_data)
        elif obj_type == SectionType.GIF89 and type(obj_data) in [bytes, bytearray]:
            self.svalues.insert(section_idx, obj_data)
        else:
            raise TypeError("Object data not valid for object type.")

        self.stypes.insert(section_idx, obj_type)
        self.nsects += 1

    def append(self, obj_type: SectionType, obj_data: Any):
        """Appends section.

        Args:
            obj_type (int): Section type of new section.
            obj_data (Any): Section value of new section.

        Example:
            >>> fpff.append(SectionType.ASCII, 'Hello, world!')
        """

        self.insert(self.nsects, obj_type, obj_data)

    def remove(self, section_idx: int):
        """Removes section at index.

        Args:
            section_idx (int): Index of section to remove.

        Example:
            >>> fpff.remove(0)
        """

        del self.svalues[section_idx]
        del self.stypes[section_idx]
        self.nsects -= 1

    def __repr__(self) -> str:
        """String representation of FPFF.

        Returns:
            string: String representation of FPFF.
        """

        return str(self.stypes)
